dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
+// ResourceClaim describes which resources (typically one or more devices)
+// are needed by a resource consumer.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClaim struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	ResourceClaimSpecAlternatives `json:",inline"` // Inlined with "spec" defined in the ResourceClaimSpec.
+
+	// Status describes whether the claim is ready for use.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ResourceClaimSpecAlternatives defines how a claim is to be allocated.
+type ResourceClaimSpecAlternatives struct {
+	// Spec define what to allocated and how to configure it.
+	// Spec and SpecRef are mutually exclusive.
+	// +optional
+	Spec *ResourceClaimSpec `json:"spec"`
+
+	// SpecRef references a separate object with the specification of the claim
+	// that will be used by the driver when allocating a resource for the
+	// claim. Parameters and ParametersRef are mutually exclusive.
+	//
+	// SpecRef is typically used to reference a vendor CR.
+	// A vendor controller then converts that CR into a ResourceClaimSpecification
+	// object and that is then used to allocate the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	// +optional
+	SpecRef *ResourceClaimSpecReference `json:"specRef,omitempty"`
 }
 
-// DeviceClaimSpec details the requirements that devices chosen
-// to satisfy this claim must meet.
-type DeviceClaimSpec struct {
-	// MatchAttributes allows specifying a constraint that will apply
-	// across all of the claims. For example, if you specified "numa", then
-	// this overall claim could only be successfully fulfilled if all of
-	// the included claims could be fulfilled by pools with the same "numa"
-	// attribute value. If we simply set matchAttributes in each claim
-	// separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// constraint.
+// Used inside a ResourceClaimSpecAlternatives or a ResourceClaimSpecification object.
+type ResourceClaimSpec struct {
+	// ControllerName defines the name of the resource driver that is
+	// meant to handle allocation of this claim. If empty,
+	// structured parameters are used for allocation.
+	//
+	// Resource drivers have a unique name in forward domain order
+	// (acme.example.com).
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	ControllerName string `json:"controllername,omitempty"`
+
+	// Config defines configuration parameters that apply to the entire claim.
+	// They are ignored while allocating the claim.
```

Comment:
@klueska had examples for why both are needed. The claim-level config was for settings affecting multiple devices while other settings were for individual devices.

Reply:
The classic example I have is with MPS -- i.e. I want to launch an MPS server that governs all of the devices allocated to this claim. Now that claims can have devices of different types (possibly even from different vendors) this line gets a bit blurry though. I'd say we still need a top-level one, I will just have to adjust what the vendor-specific API looks like so that I can reference the subset of devices I care about.